### PR TITLE
Add REACT API routes for model manager

### PR DESCRIPTION
**Pull request by lstein**
This is PR adds the following API methods for managing models:

* list_models (GET)
* update_model (PATCH)
* import_model (POST)
* delete_model (DELETE)
* convert_model (PUT)
* merge_models (PUT)

### CODE REVIEW

**Review by blessedcoolant — APPROVED** [review]

**Comment by blessedcoolant** [comment]
Is this good for a merge?

File: invokeai/app/api/routers/models.py
Comment (psychedelicious) [inline invokeai/app/api/routers/models.py:51]: This should probably be a `PATCH` if it's to change an existing entity

Comment (psychedelicious) [inline invokeai/app/api/routers/models.py:51]: This should probably be a `PATCH` if it's to change an existing entity

File: invokeai/app/api/routers/models.py
Comment (psychedelicious) [inline invokeai/app/api/routers/models.py:61]: If this is to update a model, the response model should be like `UpdateModelResponse`

File: invokeai/app/api/routers/models.py
Comment (psychedelicious) [inline invokeai/app/api/routers/models.py:73]: Doesn't the `info` include the model name? The status is implied by a 200-class response, so it should just return the updated model

I believe the convention would be to mutate the model, then retrieve it again and return the retrieved model instead.

File: invokeai/app/api/routers/models.py
Comment (psychedelicious) [inline invokeai/app/api/routers/models.py:63]: Is this the correct method? Shouldn't it be `update_model()`?

File: invokeai/app/api/routers/models.py
Comment (psychedelicious) [inline invokeai/app/api/routers/models.py:75]: Probably should have a different response type to indicate if the model update failed

File: invokeai/app/api/routers/models.py
Comment (psychedelicious) [inline invokeai/app/api/routers/models.py:67]: Per the PR description, only some of these attributes are actually changed. The `body` of the request should be typed to only include the changeable model attributes, it's very confusing if e.g. you are allowed to provide a new name, but it doesn't change anything

File: invokeai/app/api/routers/models.py
Comment (psychedelicious) [inline invokeai/app/api/routers/models.py:112]: Same as `CreateModelResponse`, we already know the name (because we called the route with it) and it's also in the info object, and the response is never sent without the status of `success`. 

If `info` is a `model info` object, I think the response model should only include that.

File: invokeai/app/api/routers/models.py
Comment (psychedelicious) [inline invokeai/app/api/routers/models.py:80]: If the `update` route is a PATCH request at `/`, does it make sense to have this `create` route be a POST request at `/`?

**Review by psychedelicious — CHANGES_REQUESTED** [review]

**Comment by psychedelicious** [comment]
I made a few changes but I'm not sure of the intent of the other items, so I've requested changes for those things. I'm happy to do the implementation with some clarification of what should happen on these routes

File: invokeai/app/api/routers/models.py
Comment (lstein) [reply invokeai/app/api/routers/models.py:51]: Done.

File: invokeai/app/api/routers/models.py
Comment (lstein) [reply invokeai/app/api/routers/models.py:61]: Done.

File: invokeai/app/api/routers/models.py
Comment (lstein) [reply invokeai/app/api/routers/models.py:63]: No, there was just the `add_model()` method with `clobber=True`. I have added an `update_model()` method to reduce confusion.

File: invokeai/app/api/routers/models.py
Comment (lstein) [reply invokeai/app/api/routers/models.py:75]: The status code has been moved out of the response object and is now is represented as various HTTP status codes (404 for model not found, 400 for invalid info)

File: invokeai/app/api/routers/models.py
Comment (lstein) [reply invokeai/app/api/routers/models.py:67]: The schema has now been fixed so that only changeable values are provided:
```
{
  "path": "string",
  "description": "string",
  "model_format": "checkpoint",
  "error": "not_found",
  "vae": "string",
  "config": "string",
  "variant": "normal"
}
```

File: invokeai/app/api/routers/models.py
Comment (lstein) [reply invokeai/app/api/routers/models.py:112]: I've removed the status code (using HTTP status instead) and renamed "name", but have not otherwise changed the `ImportModelResponse`. The reason is that there is **not** a direct relationship between the model ID provided in the call and the model information retrieved. Consider what happens when you import a civitae URL. The input is:

```
{
  "location": "https://civitai.com/api/download/models/107366"
}
```
But the output is:

```
{
  "location": "https://civitai.com/api/download/models/107366",
  "info": {
    "name": "3DMM_V12.safetensors",
    "model_type": "lora",
    "base_model": "sd-1",
    "config": {
      "path": "models/sd-1/lora/3DMM_V12.safetensors",
      "description": "sd-1 lora model 3DMM_V12",
      "model_format": "lycoris",
      "error": null
    }
  }
}
```

File: invokeai/app/api/routers/models.py
Comment (lstein) [reply invokeai/app/api/routers/models.py:80]: Yes! The only reason I chose `/import` was that I was already POSTing to `/` for the model update. But now that's been changed to PATCH, so `/` is OK.

File: invokeai/backend/install/model_install_backend.py
Comment (psychedelicious) [inline invokeai/backend/install/model_install_backend.py:177]: I get a crash on app start, as it attempts to read my `$INVOKEAI_ROOT/loras/` folder, which contains 3x `.safetensors` loras:
```py
ERROR:    Traceback (most recent call last):
  File "/home/bat/invokeai/.venv/lib/python3.10/site-packages/starlette/routing.py", line 671, in lifespan
    async with self.lifespan_context(app):
  File "/home/bat/invokeai/.venv/lib/python3.10/site-packages/starlette/routing.py", line 566, in __aenter__
    await self._router.startup()
  File "/home/bat/invokeai/.venv/lib/python3.10/site-packages/starlette/routing.py", line 648, in startup
    await handler()
  File "/home/bat/Documents/Code/InvokeAI/invokeai/app/api_app.py", line 58, in startup_event
    ApiDependencies.initialize(
  File "/home/bat/Documents/Code/InvokeAI/invokeai/app/api/dependencies.py", line 121, in initialize
    model_manager=ModelManagerService(config,logger),
  File "/home/bat/Documents/Code/InvokeAI/invokeai/app/services/model_manager_service.py", line 253, in __init__
    self.mgr = ModelManager(
  File "/home/bat/Documents/Code/InvokeAI/invokeai/backend/model_management/model_manager.py", line 348, in __init__
    self.scan_models_directory()
  File "/home/bat/Documents/Code/InvokeAI/invokeai/backend/model_management/model_manager.py", line 793, in scan_models_directory
    imported_models = self.autoimport()
  File "/home/bat/Documents/Code/InvokeAI/invokeai/backend/model_management/model_manager.py", line 849, in autoimport
    new_models_found.update(installer.heuristic_import(path))
  File "/home/bat/Documents/Code/InvokeAI/invokeai/backend/install/model_install_backend.py", line 196, in heuristic_import
    models_installed.update({str(path):self._install_path(path)})
  File "/home/bat/Documents/Code/InvokeAI/invokeai/backend/install/model_install_backend.py", line 225, in _install_path
    model_name = path.stem if info.format=='checkpoint' else path.name
AttributeError: 'NoneType' object has no attribute 'format'
```

I believe the changes to `heuristic_import()` have caused this crash

File: invokeai/app/api/routers/models.py
Comment (psychedelicious) [inline invokeai/app/api/routers/models.py:1]: In reviewing the actual objects that are returned by the endpoints (and thus comprise the OpenAPI schema) I noticed an issue. 

The list_models endpoint returns a list of `Union[tuple(OPENAPI_MODEL_CONFIGS)]` but the other endpoints return lists of `Union[tuple(MODEL_CONFIGS)]`. 

These need to be consistent and `Union[tuple(OPENAPI_MODEL_CONFIGS)]` is the right shape. I'm not sure how to fix this, I'm kinda lost in the model manager

**Review by psychedelicious — CHANGES_REQUESTED** [review]
2 changes needed - I'll note that I haven't actually tested the model manipulation functionality, only reviewing the code and types - once these changes are made, I'll create a fresh env to do functionality testing.

File: invokeai/backend/install/model_install_backend.py
Comment (lstein) [reply invokeai/backend/install/model_install_backend.py:177]: This is a separate bug that is fixed in PR #3663. However I'll port the fix over to this PR as well.

File: invokeai/app/api/routers/models.py
Comment (lstein) [reply invokeai/app/api/routers/models.py:1]: Here's what I get when I replace Union[tuple(MODEL_CONFIGS)] with Union[tuple(OPENAPI_MODEL_CONFIGS)]!
```
{
  "detail": "29 validation errors for UpdateModelResponse\ninfo -> name\n  field required (type=value_error.missing)\ninfo -> base_model\n  field required (type=value_error.missing)\ninfo -> type\n  field required (type=value_error.missing)\ninfo -> name\n  field required (type=value_error.missing)\ninfo -> base_model\n  field required (type=value_error.missing)\ninfo -> type\n  field required (type=value_error.missing)\ninfo -> model_format\n  unexpected value; permitted: <StableDiffusion1ModelFormat.Checkpoint: 'checkpoint'> (type=value_error.const; given=diffusers; permitted=(<StableDiffusion1ModelFormat.Checkpoint: 'checkpoint'>,))\ninfo -> config\n  field required (type=value_error.missing)\ninfo -> name\n  field required (type=value_error.missing)\ninfo -> base_model\n  field required (type=value_error.missing)\ninfo -> type\n  field required (type=value_error.missing)\ninfo -> name\n  field required (type=value_error.missing)\ninfo -> base_model\n  field required (type=value_error.missing)\ninfo -> type\n  field required (type=value_error.missing)\ninfo -> name\n  field required (type=value_error.missing)\ninfo -> base_model\n  field required (type=value_error.missing)\ninfo -> type\n  field required (type=value_error.missing)\ninfo -> name\n  field required (type=value_error.missing)\ninfo -> base_model\n  field required (type=value_error.missing)\ninfo -> type\n  field required (type=value_error.missing)\ninfo -> model_format\n  value is not None (type=type_error.not_none)\ninfo -> name\n  field required (type=value_error.missing)\ninfo -> base_model\n  field required (type=value_error.missing)\ninfo -> type\n  field required (type=value_error.missing)\ninfo -> model_format\n  unexpected value; permitted: <StableDiffusion2ModelFormat.Checkpoint: 'checkpoint'> (type=value_error.const; given=diffusers; permitted=(<StableDiffusion2ModelFormat.Checkpoint: 'checkpoint'>,))\ninfo -> config\n  field required (type=value_error.missing)\ninfo -> name\n  field required (type=value_error.missing)\ninfo -> base_model\n  field required (type=value_error.missing)\ninfo -> type\n  field required (type=value_error.missing)"
}
```

**Review by psychedelicious — APPROVED** [review]
Made a couple minor changes - all good! Thanks lincoln, we can move forward on the UI for this now.

**Comment by psychedelicious** [comment]
@lstein I'll defer to you to click the merge button here, not sure if there were any other internal changes needed

**Pull request merged**
